csharp/FlexVer/FlexVerComparer.cs

```diff
@@ -0,0 +1,177 @@
+	/// <returns><c>0</c> if the two versions are equal, a negative number if <c>a &lt; b</c>, or a positive number if <c>a &gt; b</c></returns>
+	public static int Compare(string a, string b) {
+		List<VersionComponent> ad = Decompose(a);
+		List<VersionComponent> bd = Decompose(b);
+		int highestCount = Math.Max(ad.Count, bd.Count);
+		for (int i = 0; i < highestCount; i++) {
+			int c = Get(ad, i).CompareTo(Get(bd, i));
+			if (c != 0) return c;
+		}
+		return 0;
+	}
+
+	private static readonly VersionComponent Null = NullVersionComponent.Instance;
+
+	internal class NullVersionComponent : VersionComponent
+	{
+		public static NullVersionComponent Instance { get; } = new();
+
+		public override int CompareTo(VersionComponent other)
+		=> ReferenceEquals(other, Null) ? 0 : -other.CompareTo(this);
+
+		private NullVersionComponent() : base(Array.Empty<int>())
+		{ }
+	}
+
+	internal class VersionComponent
+	{
+		public int[] Codepoints { get; }
+
+		public VersionComponent(int[] codepoints)
+		{
+			Codepoints = codepoints;
+		}
+
+		public virtual int CompareTo(VersionComponent that)
+		{
+			if (ReferenceEquals(that, Null)) return 1;
+
+			int[] a = this.Codepoints;
+			int[] b = that.Codepoints;
+
+			for (int i = 0; i < Math.Min(a.Length, b.Length); i++) {
+				int c1 = a[i];
+				int c2 = b[i];
+				if (c1 != c2) return c1 - c2;
+			}
+
+			return a.Length - b.Length;
+		}
+
+		public override string ToString() => new string(Codepoints.Select(el => (char)el).ToArray());
+	}
+
+	internal sealed class SemVerPrereleaseVersionComponent : VersionComponent
+	{
+		public SemVerPrereleaseVersionComponent(int[] codepoints) : base (codepoints) { }
+
+		public override int CompareTo(VersionComponent that)
+		{
+			if (ReferenceEquals(that, Null)) return -1; // opposite order
+			return base.CompareTo(that);
+		}
+	}
+
+	internal sealed class NumericVersionComponent : VersionComponent
+	{
+		public NumericVersionComponent(int[] codepoints) : base(codepoints) { }
+
+		public override int CompareTo(VersionComponent that) {
+			if (ReferenceEquals(that, Null)) return 1;
+			if (that is NumericVersionComponent) {
+				int[] a = RemoveLeadingZeroes(this.Codepoints);
+				int[] b = RemoveLeadingZeroes(that.Codepoints);
+				if (a.Length != b.Length) return a.Length-b.Length;
+				for (int i = 0; i < a.Length; i++) {
+					int ad = a[i];
+					int bd = b[i];
+					if (ad != bd) return ad-bd;
+				}
+				return 0;
+			}
+			return base.CompareTo(that);
+		}
+
+		private static int[] RemoveLeadingZeroes(int[] a) {
+			if (a.Length == 1) return a;
+			int i = 0;
+			while (i < a.Length && a[i] == '0') {
+				i++;
+			}
+			return a[i..];
+		}
+
+	}
+
+	/*
+	 * Break apart a string into intuitive version components, by splitting it where a run of
+	 * characters changes from numeric to non-numeric.
+	 */
+	internal static List<VersionComponent> Decompose(string str) {
+		// TODO: should this NRE, like in the java version, or treat `null` as `""`?
```

Comment:
Whichever feels more appropriate in C♯. I don't believe the various implementations are consistent on how they handle `null` or illegally-encoded strings — it just comes down to whatever is "most correct" in the target language.

Reply:
Sounds good. I think I'll let the public static `Compare(string, string)` require non-null inputs, but expose a default `IComparer<string>` instance that will accept null, per that interface's contract.